src/task/executable_task.rs

```diff
@@ -181,6 +197,73 @@ impl<'p> ExecutableTask<'p> {
             stderr: stderr_handle.await.unwrap(),
         })
     }
+
+    /// We store the hashes of the inputs and the outputs of the task in a file in the cache.
+    /// The current name is something like `run_environment-task_name.json`.
+    pub(crate) fn cache_name(&self) -> String {
+        format!(
+            "{}-{}.json",
+            self.run_environment.name(),
+            self.name().unwrap_or("default")
+        )
+    }
+
+    /// Checks if the task can be skipped. If the task can be skipped, it returns `CanSkip::Yes`.
+    /// If the task cannot be skipped, it returns `CanSkip::No` and includes the hash of the task
+    /// that caused the task to not be skipped - we can use this later to update the cache file quickly.
+    pub(crate) async fn can_skip(
+        &self,
+        lock_file: &LockFileDerivedData<'_>,
+    ) -> Result<CanSkip, std::io::Error> {
+        tracing::info!("Checking if task can be skipped");
+        let cache_name = self.cache_name();
+        let cache_file = self.project().task_cache_folder().join(cache_name);
+        if cache_file.exists() {
+            let cache = std::fs::read_to_string(&cache_file).unwrap();
+            let cache: TaskCache = serde_json::from_str(&cache).unwrap();
+            let hash = TaskHash::from_task(self, &lock_file.lock_file).await;
+            if let Ok(Some(hash)) = hash {
+                if hash.computation_hash() != cache.hash {
+                    return Ok(CanSkip::No(Some(hash)));
+                } else {
+                    return Ok(CanSkip::Yes);
+                }
+            }
+        }
+        Ok(CanSkip::No(None))
+    }
+
+    /// Saves the cache of the task. This function will update the cache file with the new hash of
+    /// the task (inputs and outputs). If the task has no hash, it will not save the cache.
+    pub(crate) async fn save_cache(
+        &self,
+        lock_file: &LockFileDerivedData<'_>,
+        previous_hash: Option<TaskHash>,
+    ) -> Result<(), CacheUpdateError> {
+        let task_cache_folder = self.project().task_cache_folder();
+        let cache_file = task_cache_folder.join(self.cache_name());
+        let new_hash = if let Some(mut previous_hash) = previous_hash {
+            previous_hash.update_output(self).await?;
+            previous_hash
+        } else if let Some(hash) = TaskHash::from_task(self, &lock_file.lock_file)
+            .await
+            .unwrap()
+        {
+            hash
+        } else {
+            return Ok(());
+        };
+
+        if !task_cache_folder.exists() {
+            std::fs::create_dir_all(&task_cache_folder)?;
+        }
```

Comment:
```suggestion
        std::fs::create_dir_all(&task_cache_folder)?;
```

I think you can remove the existance check because thats also done by create dir all.

Also do we want to use tokio instead of std?